File: SICP/Lecture_6_full/Code/Java/Stream/src/fayelab/sicp/stream/nostream/Procedures.java
package fayelab.sicp.stream.nostream;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;

public class Procedures
{
    public static int sumOddsSquare(Object obj)
    {
        if(obj instanceof BiTuple<?, ?>)
        {
            return sumOddsSquareForBiTuple((BiTuple<?, ?>)obj);
        }
        
        return sumOddsSquareForInteger((Integer)obj);
    }

    private static int sumOddsSquareForBiTuple(BiTuple<?, ?> biTuple)
    {
        return sumOddsSquare(biTuple.getElement1()) + sumOddsSquare(biTuple.getElement2());
    }

    private static int sumOddsSquareForInteger(Integer integer)
    {
        int n = integer.intValue();
        
        if(isOdd(n))
        {
            return square(n);
        }
        
        return 0;
    }
    
    public static List<Integer> oddFibs(int n)
    {
        List<Integer> result = new ArrayList<>();
        
        for(int i = 0; i <= n; i++)
        {
            if(isOdd(fib(i)))
            {
                result.add(i);
            }
        }
        
        return result;
    }
    
    public static List<List<Integer>> primeSumPairs(int n)
    {
        List<List<Integer>> result = new ArrayList<>();
        
        for(int i = 2; i <= n; i++)
        {
            for(int j = 1; j < i; j++)
            {
                if(isPrime(i + j))
                {
                    result.add(asList(i, j));
                }
            }
        }
        
        return result;
    }
    
    public static List<List<Integer>> triples(int n)
    {
        List<List<Integer>> result = new ArrayList<>();
        
        for(int i = 3; i <= n; i++)
        {
            for(int j = 2; j < i; j++)
            {
                for(int k = 1; k < j; k++)
                {
                    result.add(asList(i, j, k));
                }
            }
        }
        
        return result;
    }

    private static boolean isOdd(int n)
    {
        return n % 2 == 1;
    }
    
    private static int square(int n)
    {
        return n * n;
    }
    
    private static int fib(int n)
    {
        int a = 0;
        int b = 1;
        int s = 0;
        
        if(n == 0)
        {
            return a;
        }
        
        if(n == 1)
        {
            return b;
        }
        
        for(int i = 0; i <= n - 2; i++)
        {
            s = a + b;
            a = b;
            b = s;
        }
        
        return s;
    }
    
    private static boolean isPrime(int n)
    {
        for(int i = 2; i < n; i++)
        {
            if(n % i == 0)
            {
                return false;
            }
        }
        
        return true;
    }

    public static <T1, T2> BiTuple<T1, T2> biTuple(T1 element1, T2 element2)
    {
        return new BiTuple<>(element1, element2);
    }
    
    static class BiTuple<T1, T2>
    {
        private T1 element1;
        private T2 element2;
        
        public BiTuple(T1 element1, T2 element2)
        {
            this.element1 = element1;
            this.element2 = element2;
        }
    
        public T1 getElement1()
        {
            return element1;
        }
    
        public T2 getElement2()
        {
            return element2;
        }
    }
}

File: SICP/Lecture_6_full/Code/Python/nostream/procedures.py
## Procedures without stream
def sum_odds_square(obj):
    if isinstance(obj, tuple):
        return sum_odds_square(obj[0]) + sum_odds_square(obj[1])

    return square(obj) if isodd(obj) else 0

def odd_fibs(n):
    result = []
    
    for i in range(0, n + 1):
        if isodd(fib(i)):
            result.append(i)
    
    return result

def prime_sum_pairs(n):
    result = []

    for i in range(2, n + 1):
        for j in range(1, i):
            if isprime(i + j):
                result.append((i, j))

    return result

def triples(n):
    result = []

    for i in range(3, n + 1):
        for j in range(2, i):
            for k in range(1, j):
                result.append((i, j, k))

    return result

## Auxiliary functions
def isodd(n):
    return n % 2 != 0

def square(n):
    return n * n

def fib(n):
    if n == 0:
        return 0
    if n == 1:
        return 1

    return fib(n - 1) + fib(n - 2)

def isprime(n):
    for i in range(2, n):
        if n % i == 0:
            return False
    return True

## Tests
def test_sum_odds_square():
    assert 411 == sum_odds_square(((1, (2, 7)), (19, (12, 14))))
    print 'test_sum_odds_square ok'

def test_odd_fibs():
    assert [1, 2, 4, 5] == odd_fibs(6)
    print 'test_odd_fibs ok' 

def test_prime_sum_pairs():
    assert [(2, 1), (3, 2), (4, 1), (4, 3), (5, 2)] == prime_sum_pairs(5)
    print 'test_prime_sum_pairs ok'

def test_triples():
    assert [(3, 2, 1),  
            (4, 2, 1), (4, 3, 1), (4, 3, 2), 
            (5, 2, 1), (5, 3, 1), (5, 3, 2), (5, 4, 1), (5, 4, 2), (5, 4, 3)] == triples(5)
    print 'test_triples ok'

def test():
    test_sum_odds_square()
    test_odd_fibs()
    test_prime_sum_pairs()
    test_triples()

    print 'test ok'

if __name__ == '__main__':
    test()
